sharing_portal/trovi.py

```diff
@@ -0,0 +1,264 @@
+                "urn": f"urn:trovi:contents:{version.deposition_repo}:{version.deposition_id}"
+            },
+            "links": [],
+        }
+    return trovi_artifact
+
+
+def get_artifact(token, artifact_id):
+    artifact = Artifact.objects.get(pk=artifact_id)
+    url = url_with_token(f"/artifacts/{artifact.trovi_uuid}/", token)
+    res = requests.get(url)
+    check_status(res, requests.codes.ok)
+    return res.json()
+
+
+def get_artifact_by_trovi_uuid(token, artifact_id, sharing_key=None):
+    url = url_with_token(f"/artifacts/{artifact_id}/", token)
+    if sharing_key:
+        req = requests.PreparedRequest()
+        req.prepare_url(url, dict(sharing_key=sharing_key))
+        url = req.url
+    res = requests.get(url)
+    check_status(res, requests.codes.ok)
+    return res.json()
+
+
+def create_artifact(token, artifact_id):
+    artifact = Artifact.objects.get(pk=artifact_id)
+    json_data = portal_artifact_to_trovi(artifact)
+    while len(json_data["title"]) > 70:
+        print(f"This title is {len(json_data['title'])} chars (max 70)")
+        print(json_data["title"])
+        json_data["title"] = input("New title: ")
+    res = requests.post(url_with_token("/artifacts/", token), json=json_data)
+    check_status(res, requests.codes.created)
+    trovi_artifact = res.json()
+    artifact.trovi_uuid = trovi_artifact["uuid"]
+    artifact.save()
+    return trovi_artifact
+
+
+def patch_artifact(token, artifact_uuid, patches):
+    json_data = {"patch": patches}
+    print(json_data)
+    res = requests.patch(
+        url_with_token(f"/artifacts/{artifact_uuid}/", token), json=json_data
+    )
+    check_status(res, requests.codes.ok)
+    return res.json()
+
+
+def create_version(token, trovi_artifact_uuid, contents_urn, links=[]):
+    json_data = {"contents": {"urn": contents_urn}, "links": links}
+    res = requests.post(
+        url_with_token(f"/artifacts/{trovi_artifact_uuid}/versions/", token),
+        json=json_data,
+    )
+    check_status(res, requests.codes.created)
+    return res.json()
+
+
+def delete_version(token, trovi_artifact_uuid, slug):
+    res = requests.delete(
+        url_with_token(f"/artifacts/{trovi_artifact_uuid}/versions/{slug}/", token),
+    )
+    check_status(res, requests.codes.no_content)
+
+
+def increment_metric_count(token, artifact_id, version_slug, metric="access_count"):
+    res = requests.put(
+        url_with_token(
+            f"/artifacts/{artifact_id}/versions/{version_slug}/metrics?{metric}", token
+        )
+    )
+    check_status(res, requests.codes.no_content)
+
+
+def get_linked_project(artifact):
+    chameleon_projects = [
+        lp for lp in artifact["linked_projects"] if lp.split(":", 4)[3] == "chameleon"
+    ]
+    if not chameleon_projects:
+        return None
+    charge_code = chameleon_projects[0].split(":", 4)[4]
+    return Project.objects.get(charge_code=charge_code)
+
+
+def set_linked_project(artifact, charge_code, token=None):
```

Comment:
You should be able to do this in the create request if it has an admin token. Was it failing for you?